sdjournal: fix incomplete reads in JournalReader.Read()

alban opened the pull request:
This patch fixes two issues in JournalReader.Read():

1. The return value of copy() was ignored and JournalReader.Read() was
returning the wrong amount of bytes copied.

2. JournalReader implements the Reader interface so it must be able to read a
specific amount of bytes. Journald on the other hand only allows us to read
full entries of arbitrary size (without byte granularity). JournalReader
therefore needs to internally buffer entries that don't fit in the read buffer.
Callers should keep calling until 0 and/or an error is returned.

Fixes https://github.com/coreos/go-systemd/issues/172

-----


/cc @yifan-gu

alban commented:
I am testing with:
```
sudo rkt run docker://busybox --exec sh -- -c 'while sleep 1 ; do hexdump /bin/busybox ; done'
sudo rkt api-service
./client_example -follow
```

No crashes, but the client does not follow the logs correctly.

yifan-gu commented on sdjournal/read.go:
```diff
@@ -103,33 +106,42 @@ func NewJournalReader(config JournalReaderConfig) (*JournalReader, error) {
+	var sz int
+	sz, err = r.msgReader.Read(b)
+	if sz == r.msgReader.Len() {
+		r.msgReader = nil
```
@alban this seems to generate some gc pressure. Not sure if that's going to be a big deal.
Wonder if we can update the string instead of creating the new reader struct every time we read?

alban replied:
Did you notice the gc pressure? If so, how did you produce it?

alban commented:
Patch updated, PTAL.

Sometimes, I have a 30 seconds delay before getting new entries from the journal, but I think that's a problem in journald, depending on how fast it syncs the data on disk.

yifan-gu commented:
LGTM

lucab commented:
I have a probably naïf question, so sorry for asking: if this is to fix a wrong size return, wouldn't using copy()'s return-value be enough? Why introducing an additional strings.Reader? Maybe it fixes something else which is non-obvious to me...

yifan-gu commented:
@lucab because the copy() will not copy all the bytes if the provided `b []byte` is smaller than the string we read from.
Than how do we keep the remaining string?

alban commented:
I should have updated the commit message but I didn't, sorry. `r.buildMessage()` gets one full entry from journald. The cursor in journald points to an entry but is not granular to the byte so we must keep some context for the Golang Reader.

jonboulle commented:
It scares me a bit for the `JournalReader` to be maintaining this state between calls (consider `Read(); Rewind(); Read()`). At the very least I think `Read()` needs a docstring explaining very clearly that it's doing this buffering and callers must keep calling until 0 and/or an error is returned

lucab commented:
Ah, I didn't spot the change of semantics at first, my bad. It looks like this is now conflating two cases into EOF-returning (Next's EOF and log entry's EOL) but I'm not sure what (if any) will be the effect on applications.

alban commented:
I missed the issue with `Rewind()`. I updated the patch to reset the buffering in `Rewind()` and added documentation in `Read()` and updated the commit message.

@lucab I think `JournalReader.Read()` must not return EOF for the current log entry, otherwise applications would think this is an EOF for the file. It should never happen because I reset the string reader when `r.msgReader.Len() == 0`, unless we have a journal entry of zero bytes in the first place. Would it be better if I panic or just ignore the error?
```
	sz, err = r.msgReader.Read(b)
	if err != nil {
		panic("journal entry of zero bytes. This should not happen")
	}
```
or
```
	sz, _ = r.msgReader.Read(b)
```

jonboulle commented on sdjournal/read.go:
```diff
@@ -48,7 +49,8 @@ type JournalReaderConfig struct {
 // JournalReader is an io.ReadCloser which provides a simple interface for iterating through the
 // systemd journal.
 type JournalReader struct {
```
A JournalReader is not safe for concurrent use by multiple goroutines.

lucab commented on sdjournal/read.go:
```diff
@@ -101,35 +103,53 @@ func NewJournalReader(config JournalReaderConfig) (*JournalReader, error) {
+	var sz int
+	sz, err = r.msgReader.Read(b)
+	if err != nil {
+		return sz, err
```
This may return (sz > 0, EOF) once the buffer is completely consumed, at least in theory according to Reader interface.

lucab commented:
> It should never happen

I left a comment inline pinpointing what I was referring to. I'm not sure if it can really happen, but I got that impression while checking the interface doc.

> Would it be better if I panic or just ignore the error?

I generally prefer to avoid panics in libraries and have typed error wherever possible, but I know this is just my personal taste. In this case it should probably special case the EOF error condition, and clean the reader + bubble-up the error in other cases.

alban commented:
@jonboulle @lucab Thanks for the review. Patch updated.

euank commented:
This code looks dangerous and I'd like to see tests for it (though they can be in a follow up issue if needed)... but afaict this looks right and resolves the original issue, so LGTM.

lucab commented:
Also LGTM. For double-checking I added a testcase for this, up at #182; it [tested fine](https://travis-ci.org/lucab/go-systemd/builds/138901851) on travis.
